openmdao/surrogate_models/nn_interpolators/rbf_interpolator.py

```diff
@@ -0,0 +1,321 @@
+
+        dRp = frnt * np.polyval(dRp_poly, T)
+
+        # Now need dt/dx
+        xpi = np.subtract(PrdPts, self.tp[ploc[:, :-1], :])
+        xpm = PrdPts - self.tp[ploc[:, -1:], :]
+        dtx = (xpi - (T * T * xpm)) / \
+              (pdist[:, -1:, :] * pdist[:, -1:, :] * T)
+        # The gradient then is the summation across neighs of w*df/dt*dt/dx
+        grad = np.einsum('ijk,ijl...->ilk...', dRp * dtx, self.weights[ploc[:, :-1]])
+        return grad.reshape((PrdPts.shape[0], self.dep_dims, self.indep_dims))
+
+    def __init__(self, training_points, training_values, num_leaves=2, n=5, comp=2):
+        super(RBFInterpolator, self).__init__(training_points, training_values, num_leaves)
+
+        if self.ntpts < n:
+            raise ValueError('RBFInterpolator only given {0} training points, '
+                             'but requested n={1}.'
+                             .format(self.ntpts, n))
+
+        # Comp is an arbitrary value that picks a function to use
+        self.comp = comp
+
+        # For weights, first find the training points radial neighbors
+        tdist, tloc = self.KData.query(self.tp, n)
+        Tt = tdist[:, :-1] / tdist[:, -1:]
+        # Next determine weight matrix
+        Rt = self.findR(self.ntpts, Tt, tloc)
+        weights = (spsolve(csc_matrix(Rt), self.tv))[..., np.newaxis]
+
+        self.N = n
+        self.weights = weights
+
+    def __call__(self, prediction_points):
+
+        if len(prediction_points.shape) == 1:
+            # Reshape vector to n x 1 array
+            prediction_points.shape = (1, prediction_points.shape[0])
+
+        normalized_pts = (prediction_points - self.tpm) / self.tpr
+        nppts = normalized_pts.shape[0]
+        # Setup prediction points and find their radial neighbors
+        pdist, ploc = self.KData.query(normalized_pts, self.N)
+        # Check if complex step is being run
+        if (np.any(normalized_pts[0, :].imag) > 0):
+            dimdiff = np.subtract(normalized_pts.reshape((nppts, 1, self.indep_dims)),
+                                  self.tp[ploc, :])
+            # KD Tree ignores imaginary part, muse redo ndist if complex
+            pdist = np.sqrt(np.sum((dimdiff * dimdiff), axis=2))
+
+        # Take farthest distance of each point
+        Tp = pdist[:, :-1] / pdist[:, -1:]
+
+        Rp = self.findR(nppts, Tp, ploc)
+        predz = ((np.dot(Rp, self.weights[..., 0]) * self.tvr) + self.tvm).reshape(nppts, self.dep_dims)
+        return predz
+
+    def gradient(self, prediction_points):
+
+        if len(prediction_points.shape) == 1:
+            # Reshape vector to n x 1 array
+            prediction_points.shape = (1, prediction_points.shape[0])
+
+        normalized_pts = (prediction_points - self.tpm) / self.tpr
+        # Setup prediction points and find their radial neighbors
+        pdist, ploc = self.KData.query(normalized_pts, self.N)
```

Comment:
don't want to re-query here, since we don't have to. You can assume that prediction was already made at this point.